server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestGetSnapshotsAction.java

```diff
@@ -85,9 +96,52 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         getSnapshotsRequest.masterNodeTimeout(request.paramAsTime("master_timeout", getSnapshotsRequest.masterNodeTimeout()));
         return channel -> new RestCancellableNodeClient(client, request.getHttpChannel()).admin()
             .cluster()
-            .getSnapshots(
-                getSnapshotsRequest,
-                new DispatchingRestToXContentListener<>(threadPool.executor(ThreadPool.Names.MANAGEMENT), channel, request)
-            );
+            .getSnapshots(getSnapshotsRequest, new RestActionListener<>(channel) {
+                @Override
+                protected void processResponse(GetSnapshotsResponse getSnapshotsResponse) {
+                    ensureOpen();
+                    channel.sendResponse(
+                        new RestResponse(
+                            RestStatus.OK,
+                            (Objects.requireNonNullElse(request.getXContentType(), XContentType.JSON)).mediaType(),
+                            null,
+                            new ChunkedRestResponseBody() {
+
+                                final RedirectOutputStream out = new RedirectOutputStream();
+                                private XContentBuilder builder;
+
+                                private final Iterator<SnapshotInfo> snapshotInfoIterator = getSnapshotsResponse.getSnapshots().iterator();
+
+                                @Override
+                                public boolean isDone() {
+                                    return snapshotInfoIterator.hasNext() == false;
+                                }
+
+                                @Override
+                                public ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) throws IOException {
+                                    final RecyclerBytesStreamOutput chunkStream = new RecyclerBytesStreamOutput(recycler);
+                                    out.newTarget(chunkStream);
+                                    if (builder == null) {
+                                        builder = channel.newBuilder(request.getXContentType(), null, true, Streams.noCloseStream(out));
+                                        GetSnapshotsResponse.toXContentStart(builder);
+                                    }
+                                    while (snapshotInfoIterator.hasNext() && chunkStream.size() < sizeHint) {
+                                        snapshotInfoIterator.next().toXContentExternal(builder, request);
+                                    }
+                                    if (snapshotInfoIterator.hasNext() == false) {
+                                        getSnapshotsResponse.toXContentEnd(builder);
+                                        builder.close();
+                                    }
+                                    out.clearTarget();
```

Comment:
I don't like that this logic lives in the REST layer. It bypasses the `toXContent` implementation in `GetSnapshotsResponse` and I expect that over time we'll introduce bugs where these things diverge. Please could we push this logic down to the data object itself somehow (maybe giving it a different interface from `ToXContentObject`) such that the only way to build its `XContent` representation is using this chunking behaviour.

I would also find it preferable to be able to test this kind of thing without involving the REST layer.

Reply:
> It bypasses the toXContent implementation in GetSnapshotsResponse and I expect that over time we'll introduce bugs where these things diverge. Please could we push this logic down to the data object itself somehow

Sure thing, this wasn't' supposed to be the final state of things :) The plan was definitely to introduce a clean interface here and then just use that in the non-chunked `toXContent`. On it today though :)

Reply:
Sounds great (no rush tho)